File: scripts/transform/converter.py
"""페이지 타입 감지 및 메타데이터 추출.

크롤링된 원본 Markdown에서:
- 페이지 타입 감지 (api-endpoint / schema / category-index / guide)
- 제목, 설명, source URL, HTTP method/path 추출
- 출력 파일 경로(dest_path) 계산
"""

from __future__ import annotations

import re
from pathlib import Path

from scripts.core.paths import SRC_RAWS, DST_DOCS

# ── 상수 ──────────────────────────────────────────────────────────────────────

BASE_URL = "https://api.commerce.naver.com/external"

# ── 정규식 ────────────────────────────────────────────────────────────────────

METHOD_BLOCK_RE = re.compile(
    r"^```\s*\n(GET|POST|PUT|DELETE|PATCH)\s*\n\n## (/.+?)\s*\n```",
    re.MULTILINE,
)
CATEGORY_LINK_RE = re.compile(r"\[## [📄🗃️]")


# ── 페이지 타입 감지 ──────────────────────────────────────────────────────────

def detect_page_type(content: str, rel_path: str) -> str:
    """원본 Markdown 콘텐츠와 상대 경로로 페이지 타입을 감지한다.

    Returns:
        "api-endpoint" | "schema" | "category-index" | "guide"
    """
    if "schemas/" in rel_path:
        return "schema"
    if "구조체" in rel_path and rel_path.endswith(".md"):
        return "schema"
    if METHOD_BLOCK_RE.search(content):
        return "api-endpoint"
    if CATEGORY_LINK_RE.search(content):
        return "category-index"
    stem = Path(rel_path).stem
    if re.match(r"^[가-힣\-]+$", stem):
        return "category-index"
    return "guide"


# ── 메타데이터 추출 ───────────────────────────────────────────────────────────

def extract_method_and_path(content: str) -> tuple[str, str]:
    """HTTP 메서드 코드블록에서 (method, path)를 추출한다."""
    m = METHOD_BLOCK_RE.search(content)
    if m:
        return m.group(1).strip(), m.group(2).strip()
    return "", ""


def extract_source_url(content: str) -> str:
    """'> 원문: URL' 줄에서 URL을 추출한다."""
    m = re.search(r"^> 원문: (https?://\S+)", content, re.MULTILINE)
    return m.group(1).strip() if m else ""


def extract_title(content: str) -> str:
    """첫 번째 H1 제목을 추출한다."""
    m = re.search(r"^# (.+)$", content, re.MULTILINE)
    return m.group(1).strip() if m else ""


def extract_base_url(content: str) -> str:
    """curl 예제 또는 Base URL 블록에서 API base URL을 추출한다."""
    m = re.search(
        r"curl[^\n]*?'(https://[^/\s]+/[^/\s]+)(?:/|')",
        content,
    )
    if m:
        return m.group(1)
    m = re.search(r"Base URL\s*\nEdit\s*\n(https?://\S+)", content)
    if m:
        return m.group(1).strip()
    return BASE_URL


def extract_description(content: str, title: str, page_type: str) -> str:
    """첫 번째 의미 있는 단락에서 설명을 추출한다 (최대 200자)."""
    lines = content.split("\n")
    in_code = False
    candidate_lines: list[str] = []

    for line in lines:
        stripped = line.strip()
        if stripped.startswith("```"):
            in_code = not in_code
            continue
        if in_code:
            continue
        if not stripped:
            if candidate_lines:
                break
            continue
        # 노이즈 줄 스킵
        if stripped.startswith("#"):
            continue
        if stripped.startswith("> 원문:"):
            continue
        if stripped.startswith(("- curl", "- java", "- Schema", "- Example")):
            continue
        if "Direct link to" in stripped:
            continue
        if "Collapse all" in stripped or "Base URL" in stripped:
            continue
        if stripped.startswith("- application/"):
            continue
        if re.match(r"^- \d{3}$", stripped):
            continue
        if len(stripped) < 10:
            continue
        # 제목과 동일한 줄 스킵 (Docusaurus artifact)
        if stripped == title or stripped.rstrip("。.") == title.rstrip("。."):
            continue
        # 마크다운 서식 제거
        clean = re.sub(r"\*\*([^*]+)\*\*", r"\1", stripped)
        clean = re.sub(r"`([^`]+)`", r"\1", clean)
        clean = clean.replace("\\_", "_")
        candidate_lines.append(clean)
        if len(" ".join(candidate_lines)) > 120:
            break

    if candidate_lines:
        desc = " ".join(candidate_lines)
        if len(desc) > 200:
            idx = desc.rfind(".", 0, 200)
            if idx > 80:
                desc = desc[: idx + 1]
            else:
                desc = desc[:200].rstrip() + "..."
        return desc
    return title


# ── 출력 경로 계산 ────────────────────────────────────────────────────────────

def make_dest_path(
    page_type: str,
    api_path: str,
    method: str,
    src_name: str,
    *,
    dst: Path | None = None,
) -> Path:
    """변환 출력 파일의 절대 경로를 계산한다."""
    _dst = dst or DST_DOCS
    if page_type == "api-endpoint" and api_path and method:
        clean = api_path.strip("/")
        parent_parts = clean.rsplit("/", 1)
        if len(parent_parts) == 2:
            parent, last_seg = parent_parts
            return _dst / "api" / parent / f"{last_seg}.{method}.md"
        else:
            return _dst / "api" / f"{clean}.{method}.md"
    elif page_type == "schema":
        return _dst / "schema" / src_name
    elif page_type == "category-index":
        return _dst / "category" / src_name
    else:
        return _dst / "guide" / src_name


def _build_stem_dest_map(
    src: Path | None = None,
    dst: Path | None = None,
) -> dict[str, Path]:
    """소스 파일을 사전 스캔하여 {src_stem: dest_path} 매핑을 빌드한다.

    _docusaurus_url_to_rel()에서 /docs/... URL을 실제 출력 경로로 해석하는 데 사용.
    """
    _src = src or SRC_RAWS
    mapping: dict[str, Path] = {}
    for src_path in _src.rglob("*.md"):
        try:
            content = src_path.read_text(encoding="utf-8")
            rel = str(src_path.relative_to(_src)).replace("\\", "/")
            page_type = detect_page_type(content, rel)
            method, raw_api_path = extract_method_and_path(content)
            api_path = re.sub(r":([a-zA-Z][a-zA-Z0-9_]*)", r"{\1}", raw_api_path)
            dest = make_dest_path(page_type, api_path, method, src_path.name, dst=dst)
            mapping[src_path.stem] = dest
        except Exception:
            pass
    return mapping

File: src/transform/converter.ts
/**
 * 페이지 타입 감지 및 메타데이터 추출.
 */

import fs from "node:fs";
import path from "node:path";
import { resolveWritableDocsRoot, resolveWritableRawsRoot } from "../core/paths.js";

// ── 상수 ──────────────────────────────────────────────────────────────────────

export const BASE_URL = "https://api.commerce.naver.com/external";

// ── 정규식 ────────────────────────────────────────────────────────────────────

const METHOD_BLOCK_RE =
  /^```\s*\n(GET|POST|PUT|DELETE|PATCH)\s*\n\n## (\/[^\n]+?)\s*\n```/ms;
const PLAIN_METHOD_BLOCK_RE =
  /(?:^|\n)(GET|POST|PUT|DELETE|PATCH)\s*\n\s*\n## (\/[^\n]+?)\s*(?:\n|$)/m;

const CATEGORY_LINK_RE = /\[## [📄🗃️]/u;

// ── 페이지 타입 감지 ──────────────────────────────────────────────────────────

export function detectPageType(content: string, relPath: string): string {
  const decodedRelPath = safeDecode(relPath);
  if (decodedRelPath.includes("schemas/")) return "schema";
  if (decodedRelPath.includes("구조체") && decodedRelPath.endsWith(".md")) return "schema";
  if (METHOD_BLOCK_RE.test(content) || PLAIN_METHOD_BLOCK_RE.test(content)) return "api-endpoint";
  if (CATEGORY_LINK_RE.test(content)) return "category-index";
  const stem = path.basename(decodedRelPath, ".md");
  if (/^[가-힣\-]+$/.test(stem)) return "category-index";
  return "guide";
}

// ── 메타데이터 추출 ───────────────────────────────────────────────────────────

export function extractMethodAndPath(content: string): [string, string] {
  const m = content.match(METHOD_BLOCK_RE);
  if (m) return [m[1].trim(), m[2].trim()];
  const plain = content.match(PLAIN_METHOD_BLOCK_RE);
  if (plain) return [plain[1].trim(), plain[2].trim()];
  return ["", ""];
}

export function extractSourceUrl(content: string): string {
  const m = content.match(/^> 원문: (https?:\/\/\S+)/m);
  return m ? m[1].trim() : "";
}

export function extractTitle(content: string): string {
  const m = content.match(/^# (.+)$/m);
  return m ? m[1].trim() : "";
}

export function extractBaseUrl(content: string): string {
  let m = content.match(/curl[^\n]*?'(https:\/\/[^/\s]+\/[^/\s]+)(?:\/|')/);
  if (m) return m[1];
  m = content.match(/Base URL\s*\nEdit\s*\n(https?:\/\/\S+)/);
  if (m) return m[1].trim();
  return BASE_URL;
}

export function extractDescription(
  content: string,
  title: string,
  pageType: string,
): string {
  if (pageType === "category-index") {
    return extractCategoryDescription(content, title);
  }

  const lines = content.split("\n");
  let inCode = false;
  const candidateLines: string[] = [];

  for (const line of lines) {
    const stripped = line.trim();
    if (stripped.startsWith("```")) {
      inCode = !inCode;
      continue;
    }
    if (inCode) continue;
    if (!stripped) {
      if (candidateLines.length > 0) break;
      continue;
    }
    if (stripped.startsWith("#")) continue;
    if (stripped.startsWith("> 원문:")) continue;
    if (stripped.startsWith("[##")) continue;
    if (stripped.includes("](/docs/")) continue;
    if (/^- (curl|java|Schema|Example)/.test(stripped)) continue;
    if (stripped.includes("Direct link to")) continue;
    if (stripped.includes("Collapse all") || stripped.includes("Base URL")) continue;
    if (stripped.startsWith("- application/")) continue;
    if (/^- \d{3}$/.test(stripped)) continue;
    if (stripped.length < 10) continue;
    if (
      stripped === title ||
      stripped.replace(/[。.]\s*$/, "") === title.replace(/[。.]\s*$/, "")
    )
      continue;

    let clean = stripped.replace(/\*\*([^*]+)\*\*/g, "$1");
    clean = clean.replace(/`([^`]+)`/g, "$1");
    clean = clean.replace(/\\_/g, "_");
    candidateLines.push(clean);
    if (candidateLines.join(" ").length > 120) break;
  }

  if (candidateLines.length > 0) {
    let desc = candidateLines.join(" ");
    if (desc.length > 200) {
      const idx = desc.lastIndexOf(".", 200);
      if (idx > 80) {
        desc = desc.slice(0, idx + 1);
      } else {
        desc = desc.slice(0, 200).trimEnd() + "...";
      }
    }
    return desc;
  }
  return title;
}

function extractCategoryDescription(content: string, title: string): string {
  const cardTitles = [...content.matchAll(
    /\[\s*##\s+(?:[^\s]+\s+)?([^\n]+?)\s*(?:\n\s*\n[\s\S]*?)?\]\(\/docs\/[^\s)]+\)/gu,
  )].map((match) => match[1].trim());

  if (cardTitles.length === 0) {
    return `${title} 관련 API 문서를 모아둔 카테고리 인덱스입니다.`;
  }

  if (cardTitles.length === 1) {
    return `${title} 관련 API 문서를 모아둔 카테고리 인덱스입니다. 주요 문서: ${cardTitles[0]}.`;
  }

  const headline = cardTitles.slice(0, 2).join(", ");
  const extra = cardTitles.length > 2 ? ` 외 ${cardTitles.length - 2}건` : "";
  return `${title} 관련 API 문서를 모아둔 카테고리 인덱스입니다. 주요 문서: ${headline}${extra}.`;
}

// ── 출력 경로 계산 ────────────────────────────────────────────────────────────

export function makeDestPath(
  pageType: string,
  apiPath: string,
  method: string,
  srcName: string,
  opts: { dst?: string } = {},
): string {
  const dst = opts.dst ?? resolveWritableDocsRoot();
  const normalizedSrcName = safeDecode(srcName);
  if (pageType === "api-endpoint" && apiPath && method) {
    const clean = apiPath.replace(/^\//, "");
    const lastSlash = clean.lastIndexOf("/");
    if (lastSlash !== -1) {
      const parent = clean.slice(0, lastSlash);
      const lastSeg = clean.slice(lastSlash + 1);
      return path.join(dst, "api", parent, `${lastSeg}.${method}.md`);
    }
    return path.join(dst, "api", `${clean}.${method}.md`);
  }
  if (pageType === "schema") return path.join(dst, "schema", normalizedSrcName);
  if (pageType === "category-index") return path.join(dst, "category", normalizedSrcName);
  return path.join(dst, "guide", normalizedSrcName);
}

export function buildStemDestMap(
  src?: string,
  dst?: string,
): Map<string, string> {
  const _src = src ?? resolveWritableRawsRoot();
  const mapping = new Map<string, string>();

  function walk(dir: string): void {
    for (const e of fs.readdirSync(dir, { withFileTypes: true })) {
      const full = path.join(dir, e.name);
      if (e.isDirectory()) walk(full);
      else if (e.isFile() && e.name.endsWith(".md")) {
        try {
          const content = fs.readFileSync(full, "utf-8");
          const rel = path.relative(_src, full).replace(/\\/g, "/");
          const pageType = detectPageType(content, rel);
          const [method, rawApiPath] = extractMethodAndPath(content);
          const apiPath = rawApiPath.replace(/:([a-zA-Z][a-zA-Z0-9_]*)/g, "{$1}");
          const dest = makeDestPath(pageType, apiPath, method, e.name, { dst });
          mapping.set(path.basename(e.name, ".md"), dest);
        } catch {
          // ignore
        }
      }
    }
  }

  if (fs.existsSync(_src)) walk(_src);
  return mapping;
}

function safeDecode(value: string): string {
  try {
    return decodeURIComponent(value);
  } catch {
    return value;
  }
}
